### ISSUE TITLE
feature request: continue parsing ignoring hard errors inside `((..))` or add a "don't parse" mark

### ISSUE THREAD

**Issue by Kamilcuk** [koalaman/shellcheck#3105]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC1070, SC1073
- My shellcheck version (`shellcheck --version` or "online"): online, 0.10.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

There are some expressions that give shellcheck parsing errors, like:
- https://github.com/koalaman/shellcheck/issues/2947
- https://github.com/koalaman/shellcheck/issues/1291
- `for (( i = 0, max = ${#a[@]} > ${#b[@]} ? ${#a[@]} : ${#b[@]}, max > ${4:-3} ? max = ${4:-3} : 0; i < max; ++i )); do`
- `(( $invert( 1 == 1 ) ))`

#### Here's what shellcheck currently says:

```
In - line 2377:
                for (( i = 0, max = ${#a[@]} > ${#b[@]} ? ${#a[@]} : ${#b[@]}, max > ${4:-3} ? max = ${4:-3} : 0; i < max; ++i )); do
                ^-- SC1009 (info): The mentioned syntax error was in this for loop.
                    ^-- SC1073 (error): Couldn't parse this arithmetic for condition. Fix to allow more checks.
                                                                                                   ^-- SC1072 (error):  Fix any mentioned problems and try again.
```

#### Here's what I wanted or expected to see:

SC1070 and SC1073 are completely great, fine and understandable. The issue is it is __impossible__ to ignore them. Either we ignore them, in which case the rest of the file is __not__ analyzed at all, or we have a hard error forever.

It would be great if:

- there would be something like `#shellcheck parser=off` `#shellcheck parser=on` with which I would be able to disable the shellcheck parser for problematic code. 
  - which I am currently doing that with `sed '/#shellcheck parser=off/,/#shellcheck parser=on/s/.*/:/'`
- or ignoring SC1070 or SC1073 would _disable the parser_ for one expression. In this case, I would want shellcheck to just ignore everything between `((` and `))`.

Thank you, shellcheck is amazing, great project, I'm using it daily.

**Comment by wileyhy** [comment]
Just in case...  Have you tried inserting a series "disable" directives?

...
# line 2377:
# shellcheck disable=SC1009,SC1072,SC1073
for (( i = 0, max = ${#a[@]} > ${#b[@]} ? ${#a[@]} : ${#b[@]}, max >
${4:-3} ? max = ${4:-3} : 0; i < max; ++i ))
do
...

Wiley




On Sun, Dec 22, 2024 at 12:56 PM KamilCuk ***@***.***> wrote:
>
> For bugs
>
> Rule Id (if any, e.g. SC1000): SC1070, SC1073
> My shellcheck version (shellcheck --version or "online"): online, 0.10.0
> The rule's wiki page does not already cover this (e.g.
https://shellcheck.net/wiki/SC2086)
> I tried on https://www.shellcheck.net/ and verified that this is still a
problem on the latest commit
>
> For new checks and feature suggestions
>
> https://www.shellcheck.net/ (i.e. the latest commit) currently gives no
useful warnings about this
> I searched through https://github.com/koalaman/shellcheck/issues and
didn't find anything related
>
> Here's a snippet or screenshot that shows the problem:
>
> There are some expressions that give shellcheck parsing errors, like:
>
> #2947
> #1291
> for (( i = 0, max = ${#a[@]} > ${#b[@]} ? ${#a[@]} : ${#b[@]}, max >
${4:-3} ? max = ${4:-3} : 0; i < max; ++i )); do
> (( $invert( 1 == 1 ) ))
>
> Here's what shellcheck currently says:
>
> In - line 2377:
>                 for (( i = 0, max = ${#a[@]} > ${#b[@]} ? ${#a[@]} :
${#b[@]}, max > ${4:-3} ? max = ${4:-3} : 0; i < max; ++i )); do
>                 ^-- SC1009 (info): The mentioned syntax error was in this
for loop.
>                     ^-- SC1073 (error): Couldn't parse this arithmetic
for condition. Fix to allow more checks.
>
                         ^-- SC1072 (error):  Fix any mentioned problems
and try again.
>
> Here's what I wanted or expected to see:
>
> SC1070 and SC1073 are completely great, fine and understandable. The
issue is it is impossible to ignore them. Either we ignore them, in which
case the rest of the file is not analyzed at all, or we have a hard error
forever.
>
> It would be great if:
>
> there would be something like #shellcheck parser=off #shellcheck
parser=on with which I would be able to disable the shellcheck parser for
problematic code.
>
> which I am currently doing that with sed '/#shellcheck
parser=off/,/#shellcheck parser=on/s/.*/:/'
>
> or ignoring SC1070 or SC1073 would disable the parser for one expression.
In this case, I would want shellcheck to just ignore everything between ((
and )).

**Comment by Kamilcuk** [comment]
Hi @wileyhy. Yes i did try. Kindly check https://github.com/koalaman/shellcheck/issues/1291

**Comment by wileyhy** [comment]
Right. Well, here's one thing for you:  in "${4:-3}", the PE may be
somewhat ambiguous, assuming this code is written in bash. It could be read
by the parser as ${4: -3} or ${4:- 3}.

The code, of course, can be refactored to be readable by shellcheck. Some
other language other than shell  could be a more advantageous choice,
however -- perhaps a language with a few more static linters available? --
but I really don't have any way of knowing.  My experience with other
languages is limited, but in bash, at least, use of 'expr?expr:expr'
conditional operator syntax is less common.

Cheers,
Wiley
